### PR TITLE
Ordered and async publish

### LINKED ISSUE #6: Channel.basicPublish needs to be run asynchronously
The ExchangeSubscriber runs the blocking push operation on the Publisher’s thread, which is forbidden by the spec. This operation must be dispatched asynchronously, and then the request for more demand needs to be signaled afterwards from that other thread (pool).

### PR DESCRIPTION
**Pull request by LGLO**
This is to fix #6 .
I've chosen to use at most one thread*. There was STM used in other class so I used it in ExchangeSubscriber also to protect buffer of messages to send along with flag indicating if new Future should be started.

Now having buffer of messages to publish there is possibility to expose number of elements that ExchangeSubscriber requests in onSubscribe in constructor. Do you think it is useful?
Another questions is: should active Subscription be canceled on basicPublish error?

* one that can publish, more threads can pause after publishing

### CODE REVIEW

**Comment by mkiedys** [comment]
Is this tested anywhere?

**Comment by LGLO** [comment]
It was tested but not sufficiently. Today I found out that publish was called after onComplete closed channel.
Now I'm using "App" test with 3 streams:
One simply publishes N messages to to Exchange E(with routing key for Q1). Second reads from Q1 and publishes to E again with other routing key for Q2. Third reads messages from Q2 and checks if order is preserved. See https://gist.github.com/LGLO/01d31aec6fa619ed6bd4 .

I'm going to test reactive-rabbit on machine with more cores next weeks, so lets wait.

**Comment by mkiedys** [comment]
This is tricky part to get right so good test is a must.

**Comment by LGLO** [comment]
I've checked it again. Found one more rare condition: when last `basicPublish` before `onComplete` thrown then channel would have been closed twice. I also prevent closing channel that was already closed not by application. There is additional test in other PR(that build failed on Travis, but not on added test, so I guess there is some interference) that ensures messages are not lost.

File: src/main/scala/io/scalac/amqp/impl/ExchangeSubscriber.scala [hunk 1]
```diff
@@ -7,36 +7,91 @@ import com.rabbitmq.client.Channel
...
+    shutdownWhenFinished()
   }
 
   /** Our life cycle is bounded to underlying `Channel`. */
-  override def onComplete() = channel.close()
+  override def onComplete(): Unit = {
+    shutdownWhenFinished()
+  }
+
+  private def shutdownWhenFinished(): Unit = {
```
Comment (note) [inline src/main/scala/io/scalac/amqp/impl/ExchangeSubscriber.scala]: Method's body is single expression so you can omit braces (not strong opinion on that though)

File: src/main/scala/io/scalac/amqp/impl/ExchangeSubscriber.scala [hunk 2]
```diff
@@ -7,36 +7,91 @@ import com.rabbitmq.client.Channel
...
+import scala.util.control.NonFatal
 
 private[amqp] class ExchangeSubscriber(channel: Channel, exchange: String)
   extends Subscriber[Routed] {
   require(exchange.length <= 255, "exchange.length > 255")
 
   val active = new AtomicReference[Subscription]()
+  val publishingThreadRunning = Ref(false)
+  val buffer = Ref(Queue[Routed]())
+  var channelOpen = true
```
Comment (mkiedys) [inline src/main/scala/io/scalac/amqp/impl/ExchangeSubscriber.scala]: It seems that access to this variable is not synchronized anywhere. I'm right?
IMO This is more of a `closeRequested` than indicator of whether channel is closed or not.

**Comment by LGLO** [comment]
It is protected from concurrent access by `publishingThreadRunning`. It's read or written only by publishing thread or by Future that waits until publishing thread is not running.

**Comment by mkiedys** [comment]
You may not have concurrent accessed but what about visibility? It is not guaranteed that non-volatile variable will be visible across two different CPU cores - explicit memory barrier is required in order to flush the cache. Volatile annotation will generate a memory barrier.

**Comment by LGLO** [comment]
You are right. I'd like not to mix methods of handling concurrency, thus I'm wrapping this flag as `Ref[Boolean]` with your name change suggestion.

**Comment by mkiedys** [comment]
Merged here https://github.com/ScalaConsultants/reactive-rabbit/commit/09072f7528fed0569b3bc1e38c635aa3b2c00d3e.

**Pull request closed without merging**